### PR TITLE
Wowza url signing

### PR DESCRIPTION
**Pull request by jchssystems**
This is a feature to sign urls with Wowza.
To configure this, use the file configuration **org.opencastproject.security.urlsigning.provider.impl.WowzaUrlSigningProvider.cfg**

Each property has the form: key._keyId_._propertyName_

**key.wowzatoken.organization** is optional. By default is *, that means any organization.
**key.wowzatoken.url** is mandatory. describes the url prefix that urls must have.
**key.wowzatoken.secret=myTokenPrefix:mySharedSecret** is mandatory. It is a pair separated by **:**. First value, _myTokenPrefix_, means the prefix that all parameters of signed urls will have. Second value, _mySharedSecret_, defines a secret value, to secure urls.

### CODE REVIEW

**Review by rfharokh — COMMENTED** [review]
Nice of you that you finished my work: 
https://github.com/rfharokh/Opencast/commit/a253359e610a4b5caf4ce9dd2da85cd458125508

But why did you removed my name/refs?

If you need a branch based on the develop branch use this one: https://github.com/rfharokh/Opencast/tree/wowza-url-signing-feature

File: modules/urlsigning-service-impl/src/main/java/org/opencastproject/security/urlsigning/provider/impl/AbstractUrlSigningProvider.java [hunk 1]
```diff
@@ -273,7 +295,7 @@ public String sign(Policy policy) throws UrlSigningException {
       return new URI(uri.getScheme(), null, uri.getHost(), uri.getPort(), uri.getPath(),
               URLEncodedUtils.format(queryStringParameters, StandardCharsets.UTF_8), null).toString();
     } catch (Exception e) {
-      getLogger().error("Unable to create signed URL because", e);
+      getLogger().error("Unable to create signed URL because {}", ExceptionUtils.getStackTrace(e));
```
Comment (KatrinIhler) [inline modules/urlsigning-service-impl/src/main/java/org/opencastproject/security/urlsigning/provider/impl/AbstractUrlSigningProvider.java:298]: Why this change? The stack trace was logged before.

File: modules/urlsigning-service-impl/src/main/java/org/opencastproject/security/urlsigning/provider/impl/WowzaUrlSigningProvider.java [hunk 2]
```diff
@@ -46,4 +58,170 @@ public String toString() {
...
+
+    try {
+      URI uri = new URI(policy.getBaseUrl());
+
+      /*
+        For backward compatibility, but i can not see how this could work.
+        According to the documentation "https://www.wowza.com/docs/how-to-protect-streaming-using-securetoken-in-wowza-streaming-engine"
+        if you using token v1 we need a TEA implementation.
+      */
+      if ("rtmp".equals((uri.getScheme()))) {
```
Comment (KatrinIhler) [inline modules/urlsigning-service-impl/src/main/java/org/opencastproject/security/urlsigning/provider/impl/WowzaUrlSigningProvider.java:80]: RTMP was removed for Adaptive Streaming anyway a while ago (see #1179). So if this is only used for vod URLs, maybe it can be removed here, too. But I'm not sure.

File: etc/org.opencastproject.security.urlsigning.provider.impl.WowzaUrlSigningProvider.cfg [hunk 3]
```diff
@@ -25,3 +27,7 @@
 
 # key.demoKeyTwo.secret=6EDB5EDDCF994B7432C371D7C274F
```
Comment (KatrinIhler) [inline etc/org.opencastproject.security.urlsigning.provider.impl.WowzaUrlSigningProvider.cfg:28]: So does this "old" way no longer work? Then these comments should be removed.

File: modules/urlsigning-service-impl/src/main/java/org/opencastproject/security/urlsigning/WowzaResourceStrategyImpl.java [hunk 4]
```diff
@@ -39,15 +39,29 @@
   private static final String FIND_STREAM_FORMAT = ".{3}[:].*$";
   /** The URI scheme that an RTMP address uses. */
   private static final String RTMP_SCHEME = "rtmp";
+  /** The URI scheme that an http address uses. */
+  private static final String ADAPTIVESTREAMING_HTTP_SCHEME = "http";
```
Comment (KatrinIhler) [inline modules/urlsigning-service-impl/src/main/java/org/opencastproject/security/urlsigning/WowzaResourceStrategyImpl.java:43]: We removed the "adaptive" in most cases a while ago (see #1179), so just calling it streaming here would be fine.

File: docs/guides/admin/docs/configuration/stream-security.md [hunk 5]
```diff
@@ -220,6 +220,25 @@ Example:
     url.regex.archive=.*archive\/archive\/mediapackage\/.*\/.*\/.*
     url.regex.static=.*static.*
 
+Configuration of Wowza URL Signing
```
Comment (KatrinIhler) [inline docs/guides/admin/docs/configuration/stream-security.md:223]: So if I understand this correctly, this is not Stream Security, but a different way of signing URLs specifically for Wowza? (Cause I'm pretty sure the first already works.) That's bound to cause some confusion...

**Review by KatrinIhler — COMMENTED** [review]
Do I assume correctly that this only affects video-on-demand streaming and not livestreaming?

**Comment by KatrinIhler** [comment]
> Nice of you that you finished my work:
> [rfharokh@a253359](https://github.com/rfharokh/Opencast/commit/a253359e610a4b5caf4ce9dd2da85cd458125508)
> 
> But why did you removed my name/refs?

So what's up with that? The code does look pretty similar.

**Comment by gregorydlogan** [comment]
@jchssystems ping, this has been open with questions for quite a while now.

**Comment by gregorydlogan** [comment]
Between the potential copyright issues here, and the lack of progress/response we're going to close this.
